### DISCUSSION TITLE
Mui DataGrid using Zustand , cant select more then 1 row cause of unwanted Rerender

### DISCUSSION THREAD

**Discussion by eitai** [pmndrs/zustand#1397] [Q&A]
Hi, something weird happened to me when I click on a row to select its run function and pass the ID, and then filters throw an array and return the filtered values, and then set it in a Zustand state.
I don't know why but the component rerender and I couldn't select more than one each time . i debugged it and see that its happened only when the the setSelectedCustomersList zustand function runs

ZUSTAND STORE
![image](https://user-images.githubusercontent.com/35004992/198546779-2da23ccf-75bc-4ec9-b096-279dfa089749.png)

this is the function on the selected Row change 
![image](https://user-images.githubusercontent.com/35004992/198546860-38d71801-7faf-41de-8a64-12b05f938a5e.png)

and this is the Datagrid 
![image](https://user-images.githubusercontent.com/35004992/198546901-d738662a-2749-41fb-97ab-91231cd5be89.png)

thanks a lot <3

edit: added a video of what's happening
[screen-capture.webm](https://user-images.githubusercontent.com/35004992/198548340-e6079426-c480-4f17-b8a6-353188127e4d.webm)

**Comment by dai-shi** [comment]
Can anyone help?

(Maybe you want to create a very small reproduction with codesandbox.)

**Reply by eitai** [reply]
kinda hard now because its connected to my firebase

**Comment by Rec0iL99** [comment]
Hey @eitai, 

From the code that you posted, I believe you are updating the state in the wrong way.  

You need to update the state immutably.

```js
export const useCustomerListStore = create(
    persist((set, get) => ({
       customerList: [],
       selectedCustomerList: [],
       setCustomerList: (customerList) => {
           return set((state) => ({ ...state, customerList }));
       },
       setSelectedCustomersList: (selectedCustomerList) => {
           return set((state) => ({ ...state, selectedCustomerList }));
       }
    }))
);
```

I don't know if this will solve your issue, but I hope this helps.

You can refer to this [doc](https://docs.pmnd.rs/zustand/guides/immutable-state-and-merging) on how to update state the right way.

**Reply by dai-shi** [reply]
> You need to update the state immutably.

That's correct, but `...state` should be able to be omitted, because it's the default behavior.

**Reply by Rec0iL99** [reply]
Oh ok. I did not know that. 

Thanks for clarifying.
